tests/system/src/test/java/test/robot/javafx/stage/StageOwnershipTest.java

```diff
@@ -0,0 +1,438 @@
+                    assertColorEquals(COLOR7, stage6);
+                },
+                () -> assertTrue(stage6.isFocused()),
+                stage5::close,
+                () -> assertTrue(stage6.isFocused()),
+                stage6::close,
+                stage5::close,
+                () -> assertColorEquals(COLOR5, stage4));
+    }
+
+    @ParameterizedTest(name = PARAMETERIZED_TEST_DISPLAY)
+    @EnumSource(names = {"UNDECORATED", "DECORATED"})
+    void testMultiLayeredModality(StageStyle style) {
+        Util.runAndWait(() -> {
+            stage0 = createStage(style, COLOR1, null, Modality.NONE, 100, 100);
+            stage1 = createStage(style, COLOR2, stage0, Modality.WINDOW_MODAL, 150, 150);
+            stage2 = createStage(style, COLOR3, stage1, Modality.WINDOW_MODAL, 200, 200);
+
+            stage3 = createStage(style, COLOR4, null,  Modality.NONE, 600, 100);
+            stage4 = createStage(style, COLOR5, stage3,  Modality.WINDOW_MODAL, 650, 150);
+            stage5 = createStage(style, COLOR6, stage4,  Modality.WINDOW_MODAL, 700, 200);
+        });
+
+        Util.doTimeLine(WAIT_TIME,
+                stage0::show,
+                stage1::show,
+                stage2::show,
+                stage3::show,
+                stage4::show,
+                stage5::show,
+                () -> {
+                    assertColorEquals(COLOR1, stage0);
+                    assertColorEquals(COLOR2, stage1);
+                    assertColorEquals(COLOR3, stage2);
+                    assertColorEquals(COLOR4, stage3);
+                    assertColorEquals(COLOR5, stage4);
+                    assertColorEquals(COLOR6, stage5);
+                },
+                () -> assertTrue(stage5.isFocused()),
+                stage5::close,
+                () -> assertTrue(stage4.isFocused()),
+                stage4::close,
+                () -> assertTrue(stage3.isFocused()),
+                stage3::close,
+                () -> assertTrue(stage2.isFocused()),
+                stage2::close,
+                () -> assertTrue(stage1.isFocused()),
+                stage1::close,
+                () -> assertTrue(stage0.isFocused()));
+    }
+
+    @ParameterizedTest(name = PARAMETERIZED_TEST_DISPLAY)
+    @EnumSource(names = {"UNDECORATED", "DECORATED"})
+    void testIconfyRestoreChildren(StageStyle style) {
+        Util.runAndWait(() -> {
+            stage0 = createStage(style, COLOR1, null, Modality.NONE, 100, 100);
+            stage1 = createStage(style, COLOR2, stage0, Modality.WINDOW_MODAL, 150, 150);
+            stage2 = createStage(style, COLOR3, stage1, Modality.WINDOW_MODAL, 200, 200);
+        });
+
+        Util.doTimeLine(WAIT_TIME,
+                stage0::show,
+                stage1::show,
+                stage2::show,
+                () -> stage0.setIconified(true),
+                () -> {
+                    assertTrue(stage0.isIconified());
+                    assertTrue(stage1.isIconified());
```

Comment:
When a window is iconified I would expect the OS to hide any owned windows but not iconify them. For example, the owned windows should not appear as thumbnails in the dock or task bar. If we want these to be reported as iconified we would have to update glass to do that.

My first impulse is to not change anything but currently JavaFX has no way of knowing that the owned windows have changed state and can't be drawn to. We can't report them as hidden because that means something else in JavaFX. Modifying glass to report them as iconified might actually be the right thing to do even though they're not actually iconified.

Reply:
I removed the children iconified assertions and kept the color checks.